### DISCUSSION TITLE
Is there a way to generate a voxel grid?

### DISCUSSION THREAD

**Discussion by cadop** [NVIDIA/warp#74] [Q&A] [answered]
The volumes section https://nvidia.github.io/warp/_build/html/modules/runtime.html#volumes mentions loading NanoVDB and being able to access voxels. Is there a way to create this from a mesh? 

I can see there is `wp.volume_store_i` , which makes me think there might be a way to generate it myself, but even so, I would need to iterate over the mesh to do so. 

The example mesh file shows getting the closest point within radius, which is similar to what I need, but I really want all of the triangles within a radius. `bvh_query_aabb` sounds like it could be similar to a voxel query by doing aabb queries within a region, but since I don't see a size input and i dont think the bvh is going to be an even grid, i'm guessing its not what I want to use. 

I am also wondering: can a hashgrid be used as the container for a voxel volume? Or is there a more obvious way.

**Comment by mmacklin** [comment, marked as answer]
Hi @cadop,

There are a few options - you could use the `wp.bvh_query_aabb()` to get back all overlapping tris in a volume. You could compute forces from each of those (e.g.: from tri centroid / vertices).

Or you could sample the geometry with particles, and then throw those into a hash grid and use `wp.hash_grid_query()`. Using triangles means you might want to re-tessellate the mesh (similar to point sampling in some ways anyway) if you want forces to not depend too much on triangle size distribution.

Hope that helps,
Miles

**Reply by mmacklin** [reply]
There is one more option which would be something like running `wp.mesh_query_point()` multiple times. Except each time you reduce the max-distance based on the last result. So you receive successively closer points.

Not sure if that would work for your use case, but just pointing it out..

**Reply by cadop** [reply]
In the mesh query, I thought it returns the closest point? : `Computes the closest point on the mesh`.  Isn't the closest point be irrespective of the maximum distance its able to look?

**Reply by mmacklin** [reply]
Ah, sorry you're right, we'd need a min/max range for the distance.

**Comment by cadop** [comment]
So I gather that the performance difference will be a combination of triangle size + count and mesh size, and for large meshes the hashgrid will end up with some massive numbers (e.g. 50x50x50m mesh with 10cm sampling size).  Would that be a fair assumption in that circumstance to use bvh query?

**Reply by mmacklin** [reply]
Right, if you don't need uniform sampling then the BVH is likely the way to go. If you need uniform sampling then the HashGrid will probably perform better.
